File: testdata/fixtures/c/CWE-89.c
/* CWE-89: SQL Injection.
 *
 * In C the query is a char buffer and the templating is sprintf, so the
 * injection is spelled with a %s. The database sees one string and has no way
 * to know which part of it the program wrote and which part arrived from
 * outside.
 */
#include <mysql.h>
#include <sqlite3.h>
#include <stdio.h>
#include <libpq-fe.h>

void lookup(MYSQL *conn, sqlite3 *db, PGconn *pg, const char *name, int id)
{
    char query[512];
    sqlite3_stmt *stmt;

    /* ruleid: c-sql-built-by-sprintf */
    sprintf(query, "SELECT * FROM users WHERE name = '%s'", name);
    mysql_query(conn, query);

    /* ruleid: c-sql-built-by-sprintf */
    snprintf(query, sizeof query, "DELETE FROM sessions WHERE id = %d", id);
    sqlite3_exec(db, query, NULL, NULL, NULL);

    /* ruleid: c-sql-built-by-sprintf */
    snprintf(query, sizeof query, "UPDATE t SET v = '%s' WHERE k = '%s'", name, name);
    PQexec(pg, query);

    /* ruleid: c-sql-built-by-sprintf */
    sprintf(query, "insert into audit (who) values ('%s')", name);
    sqlite3_prepare_v2(db, query, -1, &stmt, NULL);

    /* ok: c-sql-built-by-sprintf */
    sqlite3_prepare_v2(db, "SELECT * FROM users WHERE name = ?", -1, &stmt, NULL);
    sqlite3_bind_text(stmt, 1, name, -1, SQLITE_STATIC);

    /* ok: c-sql-built-by-sprintf */
    mysql_query(conn, "SELECT COUNT(*) FROM users");

    char label[64];
    /* ok: c-sql-built-by-sprintf */
    snprintf(label, sizeof label, "user %s", name);
    puts(label);
}

File: testdata/fixtures/php/CWE-89.php
<?php
/* CWE-89: SQL Injection.
 *
 * PHP has no query type, so a query is a string and the templating is
 * concatenation or interpolation. Either way the database receives one string
 * and cannot tell which half the program wrote. A prepared statement is not a
 * quoting convention: the value never becomes part of the statement at all,
 * so there is no quoting to get wrong.
 */

function lookup($pdo, $mysqli, $conn, $wpdb, $name, $id)
{
    /* ruleid: php-sql-built-by-concatenation */
    $mysqli->query("SELECT * FROM users WHERE name = '" . $name . "'");
    /* ruleid: php-sql-built-by-concatenation */
    $pdo->exec("DELETE FROM sessions WHERE id = $id");
    /* ruleid: php-sql-built-by-concatenation */
    mysqli_query($conn, "UPDATE t SET v = '{$name}' WHERE k = 1");
    /* ruleid: php-sql-built-by-concatenation */
    $wpdb->get_results("SELECT * FROM wp_posts WHERE post_title = '$name'");

    $sql = "SELECT * FROM users WHERE email = '" . $_GET['email'] . "'";
    /* ruleid: php-sql-built-by-concatenation */
    $rows = $pdo->query($sql);

    /* ok: php-sql-built-by-concatenation */
    $stmt = $pdo->prepare('SELECT * FROM users WHERE name = ?');
    $stmt->execute([$name]);
    /* ok: php-sql-built-by-concatenation */
    $mysqli->query('SELECT COUNT(*) FROM users');
    /* ok: php-sql-built-by-concatenation */
    mysqli_query($conn, "SELECT id FROM users ORDER BY id");
    /* ok: php-sql-built-by-concatenation */
    $wpdb->get_results($wpdb->prepare("SELECT * FROM wp_posts WHERE ID = %d", $id));

    return [$rows, $stmt];
}
